core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommit.java

```diff
@@ -129,6 +131,19 @@ public void mutate(List<? extends Mutation> mutations) throws CrudException {
 
   @Override
   public void commit() throws CommitException, UnknownTransactionStatusException {
+    // Fill the read set with records from the write and delete sets if they are unread
+    try {
+      crud.fillReadSetForRecordsFromWriteAndDeleteSetsIfUnread();
+    } catch (CrudConflictException e) {
+      throw new CommitConflictException(
+          "Conflict occurred while reading unread records in the write and delete sets",
+          e,
+          getId());
+    } catch (CrudException e) {
+      throw new CommitException(
+          "Failed to read unread records in the write and delete sets", e, getId());
+    }
```

Comment:
Before committing (preparing) a transaction, we will fill the read set with records from the write and delete sets if they are unread.